### DISCUSSION TITLE
Setting OpenTelemetry baggage

### DISCUSSION THREAD

**Discussion by rtimush** [tokio-rs/tracing-opentelemetry#84] [Q&A]
Hi,

I'm struggling to find a way to set OpenTelemetry baggage. 

[BaggageExt](https://docs.rs/opentelemetry/0.21.0/opentelemetry/struct.Context.html#impl-BaggageExt-for-Context) trait let's me create a new Context with an updated baggage, but I haven't found a way to attach the Context to the current span. I can use `set_parent` to update the parent span context, but I haven't found a way to get the original parent context.

The following code almost works (it sets the baggage), but it makes a span reference itself as a parent, which is also wrong.
```rust
Span::current().set_parent(Span::current().context().with_baggage(baggage));
```

Do I miss something, or is it just not supported yet?

**Comment by mladedav** [comment]
There's no support for baggage in this crate.

However, from my understanding, you cannot change a span's context in OpenTelemetry which includes adding baggage. There is a context when you create the span and it is immutable so I think that the closest thing to that is what you're doing except that you actually create a new span:

```
let span = debug_span!("new_span");
span.set_parent(Span::current().context().with_baggage(baggage));
let _guard = span.entered();
```

But maybe I'm misunderstanding the otel spec.

**Reply by rtimush** [reply]
Yes, thanks, I ended up doing something like this. It is not perfect, as you have to be able to get the `Context` of the parent span, which is not always easy, but it works.

**Reply by mladedav** [reply]
I think you should be able to get the context every time, if you're creating spans with parents based on current span, you can get the current span and then its context and if you're setting parents manually, you already have the parent so you can extract the context from there.

I think the only issue would be if you're setting parents manually but only have `SpanId` and not the whole `Span` for some reason. Is that your case or is there another issue?

**Reply by rtimush** [reply]
It's easy to get the parent context if I'm the one creating the span, but not if the span is already created and entered (for example by a tower layer).

**Reply by mladedav** [reply]
I thought `Span::current().context()` would work in that case?

**Reply by rtimush** [reply]
It will give the context of the current span, while I need a context of the parent one to be able to use `set_parent`.
